### PR TITLE
Itp method

### PR DESCRIPTION
**Pull request by yash2798**
reopening the pr as the previous ci tests were failing.

Tackles [issue #184](https://github.com/SciML/NonlinearSolve.jl/issues/184)

### CODE REVIEW

**Comment by yash2798** [comment]
@ChrisRackauckas the ci tests miraculously passed here.

**Comment by yash2798** [comment]
@ChrisRackauckas This one is also ready for review in case you missed it?

File: src/itp.jl [hunk 1]
```diff
@@ -0,0 +1,105 @@
+"""
+```julia
+Itp(; k1 = Val{1}(), k2 = Val{2}(), n0 = Val{1}())
+```
+ITP (Interpolate Truncate & Project)
+
+
+"""
+
+struct Itp <: AbstractBracketingAlgorithm
```
Comment (ChrisRackauckas) [inline src/itp.jl:10]:
```suggestion
"""
```julia
Itp(; k1 = Val{1}(), k2 = Val{2}(), n0 = Val{1}())
```

Use the [ITP method](https://en.wikipedia.org/wiki/ITP_method) to find
a root of a bracketed function, with a convergence rate between 1 and 1.62.

This method was introduced in the paper "An Enhancement of the Bisection Method
Average Performance Preserving Minmax Optimality"
(https://doi.org/10.1145/3423597) by I. F. D. Oliveira and R. H. C. Takahashi.

# Tuning Parameters

The following keyword parameters are accepted.

- `n₀::Int = 1`, the 'slack'. Must not be negative.\n
  When n₀ = 0 the worst-case is identical to that of bisection,
  but increacing n₀ provides greater oppotunity for superlinearity.
- `κ₁::Float64 = 0.1`. Must not be negative.\n
  The recomended value is `0.2/(x₂ - x₁)`.
  Lower values produce tighter asymptotic behaviour, while higher values
  improve the steady-state behaviour when truncation is not helpful.
- `κ₂::Real = 2`. Must lie in [1, 1+ϕ ≈ 2.62).\n
  Higher values allow for a greater convergence rate,
  but also make the method more succeptable to worst-case performance.
  In practice, κ=1,2 seems to work well due to the computational simplicity,
  as κ₂ is used as an exponent in the method.

### Worst Case Performance

n½ + `n₀` iterations, where n½ is the number of iterations using bisection
(n½ = ⌈log2(Δx)/2`tol`⌉).

### Asymptotic Performance

If `f` is twice differentiable and the root is simple,
then with `n₀` > 0 the convergence rate is √`κ₂`.
"""
struct Itp <: AbstractBracketingAlgorithm
```

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
I made a few last edits, this looks good to go.

File: src/itp.jl [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+n½ + `n₀` iterations, where n½ is the number of iterations using bisection
+(n½ = ⌈log2(Δx)/2`tol`⌉).
+
+### Asymptotic Performance
+
+If `f` is twice differentiable and the root is simple,
+then with `n₀` > 0 the convergence rate is √`κ₂`.
+"""
+struct Itp <: AbstractBracketingAlgorithm{T}
```
Comment (ChrisRackauckas) [inline src/itp.jl:42]:
```suggestion
struct Itp{T} <: AbstractBracketingAlgorithm
```

**Comment by oscardssmith** [comment]
I realize this is post merge, but since ITP is an acronym, should it be `ITP` rather than `Itp`?

**Comment by oscardssmith** [comment]
https://github.com/SciML/SimpleNonlinearSolve.jl/pull/71

**Pull request merged**